### PR TITLE
optimizing Use expressions inside if condition

### LINKED ISSUE #111583: Collapsing two if-statements to a single if statement can result in a large performance decrease
Apologies if this has already been reported.

---

Let's say I have some code that looks like this (this is a simplified version of some code a friend was writing):

```rust
pub fn fast(mut ret: u64) -> u64 {
    let mask = (1 << 38) - 1;

    for _ in 0..100_000 {
        let mut speed = 0.0;
        let mut z: f64 = speed;
        speed += 0.200000001;

        for _ in 2..14 {
            z += speed;

            if (z.to_bits() >> 8) & mask == 0 {
                if z % 0.0625 < 1e-13 {
                    println!("{}", z % 0.0625);
                    ret += 1;
                }
            }
        }
    }

    eprintln!("ret: {ret}");
    ret
}
```

I might be tempted to collapse the if-statement in the middle, since it shouldn't change anything - in fact, clippy will recommend that I do so:

```rust
pub fn slow(mut ret: u64) -> u64 {
    let mask = (1 << 38) - 1;

    for _ in 0..100_000 {
        let mut speed = 0.0;
        let mut z: f64 = speed;
        speed += 0.200000001;

        for _ in 2..14 {
            z += speed;

            if (z.to_bits() >> 8) & mask == 0 && z % 0.0625 < 1e-13 {
                println!("{}", z % 0.0625);
                ret += 1;
            }
        }
    }

    eprintln!("ret: {ret}");
    ret
}
```

However, if I pit these two against each other using criterion, then when I run a bench (on 1.69.0):

```bash
➜ cargo bench 2> out.txt

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s


running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

slow                    time:   [7.5115 ms 7.5313 ms 7.5583 ms]
Found 4 outliers among 100 measurements (4.00%)
  2 (2.00%) high mild
  2 (2.00%) high severe

fast                    time:   [577.02 µs 578.91 µs 581.29 µs]
Found 5 outliers among 100 measurements (5.00%)
  1 (1.00%) high mild
  4 (4.00%) high severe
```

For some reason, collapsing the if branch leads to a massive performance regression! This is surprising as well since from my testing, where I set `z = 0`, the if branch should never run. Putting the two bits of code on Godbolt seems to also show that there's a bit of a difference in terms of assembly generation ([fast](https://godbolt.org/z/n969brv1b), [slow](https://godbolt.org/z/5M5fc3ae3)).

Furthermore, from some testing, commenting out either the `eprintln` or the `println` on both would result in them having similar performance.

I can set up a repo with my exact setup if that will be helpful.

### PR DESCRIPTION
**Pull request by azizghuloum**
First time contributor here.

While debugging the issue, I noticed that the HIR tree for the "slow" part contained `Use` expressions that were not handled in any special way during lowering to MIR, and thus introduced  a temporary variable and lead to the performance degradation reported.

The change of course affects how MIR is generated, and thus affects a number of tests.  In particular, it makes one (ui) test pass when it used to compile with error.  The other tests affected, according to my (shallow) analysis, look correct.

Feedback appreciated.

Fixes #111583

### CODE REVIEW

**Comment by petrochenkov** [comment]
r? rust-lang/compiler

**Comment by cjgillot** [comment]
Could you add a reduced version of the code in https://github.com/rust-lang/rust/issues/111583 as a mir-opt test? In the `tests/mir-opt/pre-codegen` directory there are several examples.

I'm a bit worried by this change. `ExprKind::Use` is responsible for dropping temporaries. The ui test changing is probably a symptom. Could you elaborate why this change is sound?

cc @dingxiangfei2009 as have better knowledge of MIR scoping than I

**Comment by dingxiangfei2009** [comment]
@azizghuloum Thank you for your PR! I would like to supply additional information for the review.

The purpose of `ExprKind::Use` is indeed for scoping in MIR. It is used for inserting drops while building MIRs, but it is an indirect effect because its sole purpose is a "marker" to insert instructions to drop temporaries while evaluating the `source`.

By bypassing the `as_temp` step, the temporaries are then dropped at a later program point because the effective terminating scope of them are higher than we would like it to be. Maybe our current test suite has not tested this yet. That could be a source of problem.

There are two interesting questions remained. One is how shall we also admin the program in https://github.com/rust-lang/rust/pull/111644#issuecomment-1551092311 and another is why the performance degradation happens. I would need some extra time on the investigation. I will be right back.

**Comment by azizghuloum** [comment]
Working on another solution which does not affect lifetimes... testing now.

**Comment by dingxiangfei2009** [comment]
@azizghuloum I digged deeper into the code today and I would like to say that I am glad that you have kept this PR open. I believe it is still safe to unwrap the `ExprKind::Use` here. The reason is that the scope information has been computed in the lowering to THIR, which I overlooked. From now on, `as_temp` and eventually `as_place` will consult with the `temp_lifetime` scoping information to insert the drops.

cc @cjgillot 

Also, the reason that employing `let_chains` makes the example compile is because let-chains are handled completely differently. Chaining lets can only be written and connected by `&&`. Exploiting this feature, the match failure cases are always routed to breaking into the `else` block. For example, the version with the let-chain gets compiled into this:
g2
Without let-chain and without your patch, it would look like this:
g
Now you would see that, the control flow would always take on a path on which `_1` is initialized. However, this fact can only be exploited after running the `ConstProp` and `SimplifyConstCondition` MIR optimization passes which are applied only after the MIR borrow checking. Your patch, however, would successfully mitigate this issue in my humble opinion. :+1:

**Comment by cjgillot** [comment]
@azizghuloum like I asked in https://github.com/rust-lang/rust/pull/111752, could you add a test for this new behaviour in the mir-opt/built directory? There are examples in that directory.

That test should involve each term in the logical expression to have stuff to drop (`struct Droppy;` with an empty `Drop` impl should be enough), so we can see that this remains correct.

This PR makes new code compile, so we'll need a signoff from t-lang.

**Comment by azizghuloum** [comment]
@cjgillot is this the test you were looking for to ensure drops are handled properly?  I'll add another test from issue #111583 .

**Comment by azizghuloum** [comment]
@cjgillot what's a _"signoff from t-lang"_?  Sorry, new here.

File: tests/mir-opt/building/logical_and_in_conditional.rs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        let _t = Droppy;
+        a == 0
+    } && {
+        let _t = Droppy;
+        b == 0
+    } && {
+        let _t = Droppy;
+        c == 0
+    } {}
+}
```
Comment (cjgillot) [inline tests/mir-opt/building/logical_and_in_conditional.rs:21]:
```suggestion
struct Droppy(bool);

impl Drop for Droppy {
    fn drop(&mut self) {
        println!("drop");
    }
}

// EMIT_MIR logical_and_in_conditional.test_and.built.after.mir
fn test_and(a: i32, b: i32, c: i32) {
    if Droppy(a).0 && Droppy(b).0 && Droppy(c).0 {}
}
```

Thanks. Nit: curly braces create their own scope. We want to ensure that they are dropped in the implicit scope of the condition.

**Comment by azizghuloum** [comment]
Did add the test, and as I expected, there are now two drop calls for each `Droppy` (one drop at top of the consequent side and one at the top of the alternate side).  Dunno if this is a problem but that's why I was going for the alternative solution at #111725.

**Comment by azizghuloum** [comment]
Greetings again,

I have fixed PR #111725 to

1. perform the desired optimization but at AST->HIR lowering level (instead of HIR->MIR lowering which this one does)
2. not skip over the `Use` expressions (which this one does)
3. not emit duplicate `drop` calls (which this one does)
4. have the two new tests that I've added here.

Can we review that one and close this PR please? @cjgillot @dingxiangfei2009

**Comment by cjgillot** [comment]
Replaced by https://github.com/rust-lang/rust/pull/111725

**Pull request closed without merging**